src/java.base/share/classes/java/io/OutputStream.java

```diff
@@ -204,4 +204,15 @@ public void flush() throws IOException {
     public void close() throws IOException {
     }
 
+    /**
+     * Checks whether this OutputStream is trusted, i.e. doesn't modify written byte[]
+     *
+     * @return true if the argument of {@link #write(byte[])}} and {@link #write(byte[], int, int)}} needn't be copied
+     */
+    boolean trusted() {
```

Comment:
Technically speaking, `OutputStream` is an `abstract class`, so this declaration of `boolean trusted()` is a package-protected method that will be visible and overridable only within JDK itself.
However, I agree it looks suspicious.

Reply:
The comment on this method doesn't look right. The issue for BAIS is that its lifetime may be different to the byte[] that it wraps, think about use after the BAIS has been discarded. You don't want a sink keeping a reference to the byte[] even if it doesn't scribble on it.

Reply:
@mkarg I guess the method can only be implemented only by subclasses residing in the same package with `OutputStream`, right?

Reply:
@stsypanov Yes but still it is just weird to ask any output stream if *it* is trusted. I mean, it feels just unsecure to ask: "Do *you* pretend to be trusted?" instead of "Do *we* trust you?". I could sleep better if this method would not be part of each OutputStream subclass. We should either move it back to the place where needed, or into some static utility like `OutputStreams::isTrusted(OutputStream)` (mind the plural!), or it should at least be `final`.

Reply:
And it should not only talk about modifying the buffer. As it’s also about leaking the reference.

Reply:
Unlike BAIS, the BufferedInputStream can wrap an untrusted InputStream.  BIS.buf is passed directly to wrapped InputStream so I would assume that we would want to avoid exposing BIS.buf to the `out` parameter of transferTo.  This way we know the input stream is not able to poison the output stream when a write is in progress.  

I assume that the current small list of trusted classes are also immune to poison so I imagine this patch is safe.  However, for any FilterInputStream we should either always use Arrays.copyOfRange because the input side can poison the output side or it needs a mirroring allow list for the target input stream to insure that the wrapped input stream is not poisoning the out parameter.

For instance, java.util.zip.CheckedOutputStream in theory could be added as trusted class as it doesn't leak or poison but, looking at the code it would appear that it is not immune to poison.